server/jetstream_api.go

```diff
@@ -890,9 +890,30 @@ func (js *jetStream) apiDispatch(sub *subscription, c *client, acc *Account, sub
 	// Copy the state. Note the JSAPI only uses the hdr index to piece apart the
 	// header from the msg body. No other references are needed.
 	// Check pending and warn if getting backed up.
+retry:
 	pending, _ := s.jsAPIRoutedReqs.push(&jsAPIRoutedReq{jsub, sub, acc, subject, reply, copyBytes(rmsg), c.pa})
 	limit := atomic.LoadInt64(&js.queueLimit)
 	if pending >= int(limit) {
+		if _, ok := s.jsAPIRoutedReqs.popOne(); ok {
+			// If we were able to take one of the oldest items off the queue, then
```

Comment:
I think the code is no longer properly updating the `js.apiInflight` that is bumped above and is normally decremented when processing elements in `processJSAPIRoutedRequests()`.

You have introduced quite recently ability to limit the ipQueue with making the push() fail when above that limit, why not using this instead of some arbitrary limit leading to a drain?

Reply:
Have updated the code for `apiInflight`. I think the original reason for not using the `push()` error on limits was that it would make it impossible to make the queue limit reloadable without rebuilding the IPQ. Right now it isn't reloadable anyway, but today we could feasibly fix that.

Reply:
ISWYM, have pushed another fix.